apps/openmw/mwmechanics/activespells.cpp

```diff
@@ -165,14 +160,44 @@ namespace MWMechanics
         params.mDisplayName = displayName;
         params.mCasterActorId = casterActorId;
 
-        if (!exists || stack)
-            mSpells.insert (std::make_pair(id, params));
+        if (it == end() || stack)
+        {
+            mSpells.insert(std::make_pair(id, params));
+        }
         else
-            mSpells.find(id)->second = params;
+        {
+            // addSpell() is called with effects for a range.
+            // but a spell may have effects with different ranges (e.g. Touch & Target)
+            // so, if we see new effects for same spell assume additional 
+            // spell effects and add to existing effects of spell
+            mergeEffects(params.mEffects, it->second.mEffects);
+            it->second = params;
+        }
 
         mSpellsChanged = true;
     }
 
+    void ActiveSpells::mergeEffects(std::vector<ActiveEffect>& addTo, const std::vector<ActiveEffect>& from)
+    {
+        for (std::vector<ActiveEffect>::const_iterator effect(from.begin()); effect != from.end(); ++effect)
+        {
+            // if effect is not in addTo, add it
+            bool missing = true;
+            for (std::vector<ActiveEffect>::const_iterator iter(addTo.begin()); iter != addTo.end(); ++iter)
+            {
+                if (effect->mEffectId == iter->mEffectId)
```

Comment:
This would be slightly incorrect when there's skills/attributes involved. Try effect->mEffectId == iter->mEffectId && effect->mArg == iter->mArg.